modules/contacts/modules.php

```diff
@@ -84,6 +84,31 @@ public function process() {
     }
 }
 
+/**
+ * @subpackage contacts/handler
+ */
+class Hm_Handler_process_export_contacts extends Hm_Handler_Module {
+    public function process() {
+        if (array_key_exists('contact_source', $this->request->get)) {
+            $source = $this->request->get['contact_source'];
+            $contact_list = $this->user_config->get('contacts', array());
+            if ($source != 'all') {
+                $contact_list = array_filter($contact_list, function($v) { return $v['source'] == $this->request->get['contact_source']; });
+            }
+
+            header('Content-Type: text/csv');
+            header('Content-Disposition: attachment; filename="'.$source.'_contacts.csv"');
+            $output = fopen('php://output', 'w');
+            fputcsv($output, array('display_name', 'email_address', 'phone_number'));
+            foreach ($contact_list as $contact) {
+                fputcsv($output, array($contact['display_name'], $contact['email_address'], $contact['phone_number']));
+            }
+            fclose($output);
+            exit;
```

Comment:
I think we should be more creative here and disallow direct output from handler modules and stopping the whole application in the middle of execution. It violates the purpose of these modules and also prevents us from properly debugging in the future. It also skips sometimes other important application code from executing if you die/exit early.

If you check lib/dispatch.php you will see that normal requests are dispatched via Hm_Dispatch->process_request which calls render_output at the end. It assembles a Hm_Output_HTTP class which renders the headers and then sends the headers to the browser. This is exactly what we want. So, instead of doing it in the handler, you can generate the csv as string (or store as array of values), pass that to output modules, have a special output module that formats the array as csv and renders that as a single output content. Then, the Hm_Output_HTTP class will handle the rest. So, it is important that the new output module render only the csv string, pass the needed headers and prevent any other output from being displayed. We might need to tweak Hm_Output_Module classes to allow a particular child module to reset the output and replace with a particular string that will be the only output sent to the browser. This will be useful when we want to stream contents of files, download other types of content from within cypht itself as some future tasks.